prov/verbs/src/fi_verbs.c

```diff
@@ -747,124 +747,63 @@ static int fi_ibv_msg_ep_bind(struct fid *fid, struct fid *bfid, uint64_t flags)
-	wr.next = NULL;
-	wr.sg_list = &sge;
-	wr.num_sge = 1;
-
-	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
-	return -ibv_post_recv(_ep->id->qp, &wr, &bad);
-}
-
-static ssize_t
-fi_ibv_msg_ep_recvv(struct fid_ep *ep, const struct iovec *iov, void **desc,
-                 size_t count, fi_addr_t src_addr, void *context)
+fi_ibv_msg_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flags)
 {
 	struct fi_ibv_msg_ep *_ep;
 	struct ibv_recv_wr wr, *bad;
-	struct ibv_sge *sge;
+	struct ibv_sge *sge = NULL;
 	size_t i;
 
-	sge = alloca(count * sizeof(struct ibv_sge));
-	wr.wr_id = (uintptr_t) context;
+	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
+	wr.wr_id = (uintptr_t) msg->context;
 	wr.next = NULL;
-	wr.sg_list = sge;
-	wr.num_sge = (int) count;
+	if (msg->iov_count) {
+		sge = alloca(sizeof(*sge) * msg->iov_count);
+		for (i = 0; i < msg->iov_count; i++) {
+			sge[i].addr = (uintptr_t) msg->msg_iov[i].iov_base;
+			sge[i].length = (uint32_t) msg->msg_iov[i].iov_len;
+			sge[i].lkey = (uint32_t) (uintptr_t) (msg->desc[i]);
+		}
 
-	for (i = 0; i < count; i++) {
-		sge[i].addr = (uintptr_t) iov[i].iov_base;
-		sge[i].length = (uint32_t) iov[i].iov_len;
-		sge[i].lkey = (uint32_t) (uintptr_t) desc[i];
 	}
+	wr.sg_list = sge;
+	wr.num_sge = msg->iov_count;
 
-	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
 	return -ibv_post_recv(_ep->id->qp, &wr, &bad);
 }
 
 static ssize_t
-fi_ibv_msg_ep_send(struct fid_ep *ep, const void *buf, size_t len,
-		void *desc, fi_addr_t dest_addr, void *context)
-{
-	struct fi_ibv_msg_ep *_ep;
-	struct ibv_send_wr wr, *bad;
-	struct ibv_sge sge;
-
-	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
-	sge.addr = (uintptr_t) buf;
-	sge.length = (uint32_t) len;
-	sge.lkey = (uint32_t) (uintptr_t) desc;
-
-	wr.wr_id = (uintptr_t) context;
-	wr.next = NULL;
-	wr.sg_list = &sge;
-	wr.num_sge = 1;
-	wr.opcode = IBV_WR_SEND;
-	wr.send_flags = (len <= _ep->inline_size) ? IBV_SEND_INLINE : 0;
-
-	return -ibv_post_send(_ep->id->qp, &wr, &bad);
-}
-
-static ssize_t
-fi_ibv_msg_ep_senddata(struct fid_ep *ep, const void *buf, size_t len,
-		    void *desc, uint64_t data, fi_addr_t dest_addr, void *context)
+fi_ibv_msg_ep_recv(struct fid_ep *ep, void *buf, size_t len,
+		void *desc, fi_addr_t src_addr, void *context)
 {
-	struct fi_ibv_msg_ep *_ep;
-	struct ibv_send_wr wr, *bad;
-	struct ibv_sge sge;
-
-	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
-	sge.addr = (uintptr_t) buf;
-	sge.length = (uint32_t) len;
-	sge.lkey = (uint32_t) (uintptr_t) desc;
-
-	wr.wr_id = (uintptr_t) context;
-	wr.next = NULL;
-	wr.sg_list = &sge;
-	wr.num_sge = 1;
-	wr.opcode = IBV_WR_SEND_WITH_IMM;
-	wr.send_flags = (len <= _ep->inline_size) ? IBV_SEND_INLINE : 0;
-	wr.imm_data = (uint32_t) data;
+	const struct iovec iov = {
+		.iov_base = (void *)buf,
+		.iov_len = len,
+	};
+	struct fi_msg msg = {
+		.msg_iov = &iov,
+		.desc = &desc,
+		.iov_count = 1,
+		.addr = src_addr,
+		.context = context,
+	};
```

Comment:
I don't think this type of initialization is standard within a function.  I know I've seen some compilers complain about it in the past.